cmd/kube-proxy/app/server.go

```diff
@@ -228,6 +229,8 @@ func (o *Options) Complete() error {
 		if err != nil {
 			return err
 		}
+		// make flags take precedence
+		mergo.Merge(c, o.config, mergo.WithOverride)
```

Comment:
does this make defaulted values override explicit config values?

Reply:
No, have a test case here: https://github.com/kubernetes/kubernetes/blob/2e112f2efe15907af9b2dc2e485e74bc980df429/cmd/kube-proxy/app/server_test.go#L663-L668
You can see if the flag is not set but proxy component config is set, after the merge the value will be the one in proxy component config.

Reply:
the unit test that makes sure this works doesn't match how Options are actually used by kube-proxy

kube-proxy:
1. creates options: `opts := NewOptions()`
2. applies defaults: `opts.config, err = opts.ApplyDefaults(opts.config)`
3. binds flags

I'm pretty sure the defaults in step 2 will stomp any config value specified with this current PR.

In the unit test, make sure you are exercising this the same way, and make sure the config file is effective when setting a value that is defaulted by opts.ApplyDefaults

Reply:
> I'm pretty sure the defaults in step 2 will stomp any config value specified with this current PR.

If that turns out to be the case, that means our testing of the config file path has gaps that need to be closed as well

Reply:
I see, that's a good point. I think the tricky part is it's hard to determine whether the value is from the default or explicitly passed from flags. The merge logic here is to make the values in `o.config` take precedence if it's a non-empty value, which can be our defaults.
I think one way to fix it is to move all default values for flags to a default proxy config struct, and merge it later without overwriting.

Reply:
> I think one way to fix it is to move all default values for flags to a default proxy config struct, and merge it later without overwriting.

That makes the help displayed for the flags much less helpful, since it doesn't show the defaults used if no flags are passed. Before restructuring defaulting, we should determine whether this change is the right direction, see the comment in https://github.com/kubernetes/kubernetes/pull/101273#issuecomment-823357862